src/lib/qgram.rs

```diff
@@ -0,0 +1,235 @@
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+#![warn(missing_docs)]
+
+/// Calculate a q-gram distance between strings. Implementation based on the
+/// simmetrics library: https://github.com/Simmetrics/simmetrics
+
+use std::hash::Hash;
+
+use multiset::HashMultiSet;
+
+/// q-grams are padded at the start of the string, q-1 times.
+///
+/// E.g. if `q` is 3, the string "ABC" becomes "##ABC" with the default character.
+const DEFAULT_START_PADDING: char = '#';
+
+/// q-grams are padded at the end of the string, q-1 times.
+///
+/// E.g. if `q` is 3, the string "ABC" becomes "ABC##" with the default character.
+const DEFAULT_END_PADDING: char = '#';
+
+struct QGram {
+    q: u32,
+    filter: bool,
+    start_padding: String,
+    end_padding: String,
+}
+
+impl QGram {
+    fn new(q: u32, start_padding: String, end_padding: String) -> QGram {
+        QGram { q,
+                filter: false,
+                start_padding,
+                end_padding, }
+    }
+
+    fn compare<T: Clone + Eq + Hash>(&self, a: &HashMultiSet<T>, b: &HashMultiSet<T>) -> f64 {
+        if a.is_empty() && b.is_empty() {
+            return 1.0;
+        }
+        if a.is_empty() || b.is_empty() {
+            return 0.0;
+        }
+        let dist = self.distance(a, b);
+        1.0 - dist / f64::from(a.len() as i32 + b.len() as i32)
+    }
+
+    fn distance<T: Clone + Eq + Hash>(&self, a: &HashMultiSet<T>, b: &HashMultiSet<T>) -> f64 {
+        let mut distance: f64 = 0.0;
+        let union = a.clone() + b.clone();
+        for token in union.distinct_elements() {
+            distance += f64::from((a.count_of(token) as i32 - b.count_of(token) as i32).abs());
+        }
+        distance
+    }
+
+    /// Parse a string into a multiset of q-grams.
+    fn tokenise(&self, input: &str) -> HashMultiSet<String> {
+        let mut token_set = HashMultiSet::new();
+        if input.is_empty() {
+            return token_set;
+        }
+        let mut padded = self.start_padding.clone();
+        padded.push_str(input);
+        padded.push_str(&self.end_padding);
+        // A string must contain at least `self.q` characters.
+        if self.filter && padded.len() < self.q as usize {
+            return token_set;
+        } else if padded.len() < self.q as usize {
+            token_set.insert(padded);
+            return token_set;
+        }
+        // To parse a string into a q-gram set of tokens we move a `self.q`-wide
+        // sliding windows across the string.
+        let chars = padded.chars().collect::<Vec<char>>();
+        let last_qgram_index = chars.len() - self.q as usize;
+        for qgram_start in 0..=last_qgram_index {
```

Comment:
Stray `=`? Or maybe it does something I don't know about?

Reply:
This is equivalent to `for (qgram_start = 0; i <= last_qgram_index; i++) { ... }` so the `..=` is like the Java `<=`.

Reply:
I didn't know that. I don't like it either (Rust, why so much syntax?!) but fair enough.

Reply:
Gross!